src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -431,11 +442,15 @@ internal void CompleteReader(Exception exception)
 
             lock (_sync)
             {
+                // If we're reading, treat clean up that state before continuting
                 if (_readingState.IsActive)
                 {
-                    ThrowHelper.ThrowInvalidOperationException_CompleteReaderActiveReader();
+                    _readingState.End();
                 }
 
+                // REVIEW: We should consider cleaning up all of the allocated memory
```

Comment:
I was intending to do something like this for a long time, the biggest problem is the producer that may continue writing after reader completed, so we don't want to return everything but keep a single segment around to return to the writer over and over again.

Reply:
How do we currently handle backpressure after reader completion?

Reply:
We signal the writer with IsCompleted true when the reader completes.

Reply:
What if it keeps writing? We'll just expand the pipe forever?

Reply:
Yes. That's the case today right? This doesn't fix any problems there. I originally made this thing auto advance but there's no point because if you do call advance and you consume nothing, you're in the same boat.

This line of code is supposed to handle the case where the reader is completed:

https://github.com/davidfowl/corefx/blob/4dcaa60893740a2c7a934af9185d6a5cff62699b/src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L238

I'll add a test that does the following:
1. Writes until there's backpressure
1. Complete the reader
1. Write again, awaitable returned should be completed since there's nobody to drain.